Sources/PromotedCore/MetricsLogger/MetricsLogger+Validation.swift

```diff
@@ -0,0 +1,59 @@
+import Foundation
+import SwiftProtobuf
+
+extension MetricsLogger {
+  func validate(message: Message) {
+    switch message {
+    case let logRequest as Event_LogRequest:
+      validate(logRequest: logRequest)
+    case let impression as Event_Impression:
+      validate(impression: impression)
+    case let action as Event_Action:
+      validate(action: action)
+    case let user as Event_User:
+      validate(user: user)
+    default:
+      break
+    }
+  }
+
+  private func validate(logRequest: Event_LogRequest) {
+    if logRequest.userInfo.logUserID.isEmptyOrWhitespace {
+      handleLoggingError(.missingLogUserIDInLogRequest)
+    }
+  }
+
+  private func validate(impression: Event_Impression) {
+    if (
+      impression.sourceType == .delivery &&
+      impression.insertionID.isEmptyOrWhitespace &&
+      impression.contentID.isEmptyOrWhitespace
+    ) {
+      handleLoggingError(.missingJoinableFieldsInImpression)
+    }
+  }
+
+  private func validate(action: Event_Action) {
+    if (
+      action.actionType != .checkout &&
+      action.actionType != .purchase &&
+      action.impressionID.isEmptyOrWhitespace &&
+      action.insertionID.isEmptyOrWhitespace &&
+      action.contentID.isEmptyOrWhitespace
+    ) {
+      handleLoggingError(.missingJoinableFieldsInAction)
+    }
+  }
+
+  private func validate(user: Event_User) {
+    if user.userInfo.logUserID.isEmpty {
```

Comment:
Snit - I think you also want to check userID too.

Best practice wise - we'll want to log this from the server-side.

Reply:
Do you mean:
```swift
  private func validate(user: Event_User) {
    if (
      user.userInfo.logUserID.isEmptyOrWhitespace &&
      user.userInfo.userID.isEmptyOrWhitespace
    ) {
      handleLoggingError(.missingLogUserIDInUserMessage)
    }
  }
```